src/Types/InputType.php

```diff
@@ -0,0 +1,127 @@
+     */
+    public function __construct(string $className, string $inputName, ?string $description, bool $isUpdate, FieldsBuilder $fieldsBuilder)
+    {
+        $this->fields = $fieldsBuilder->getInputFields($className, $inputName, $isUpdate);
+
+        $fields = function() use ($isUpdate) {
+            $fields = [];
+            foreach ($this->fields as $name => $field) {
+                $type = $field->getType();
+
+                $fields[$name] = [
+                    'type' => $type,
+                    'description' => $field->getDescription(),
+                ];
+
+                if ($field->hasDefaultValue() && !$isUpdate) {
+                    $fields[$name]['defaultValue'] = $field->getDefaultValue();
+                }
+            }
+
+            return $fields;
+        };
+
+        $config = [
+            'name' => $inputName,
+            'description' => $description,
+            'fields' => $fields,
+        ];
+
+        parent::__construct($config);
+        $this->className = $className;
+    }
+
+
+    /**
+     * @param object|null          $source
+     * @param array<string, mixed> $args
+     * @param mixed                $context
+     * @param ResolveInfo          $resolveInfo
+     *
+     * @return object
+     */
+    public function resolve(?object $source, array $args, $context, ResolveInfo $resolveInfo): object
+    {
+        $mappedValues = [];
+        foreach ($this->fields as $field) {
+            $name = $field->getName();
+            if (array_key_exists($name, $args)) {
+                $mappedValues[$field->getPropertyName()] = $field->resolve($source, $args, $context, $resolveInfo);
+            }
+        }
+
+        $instance = $this->createInstance($mappedValues);
+        foreach ($mappedValues as $property => $value) {
+            PropertyAccessor::setValue($instance, $property, $value);
+        }
+
+        return $instance;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function decorate(callable $decorator): void
+    {
+        throw FailedResolvingInputType::createForDecorator();
+    }
+
+    /**
+     * Creates an instance of the input class.
+     *
+     * @param array<string, mixed> $values
+     *
+     * @return object
+     */
+    private function createInstance(array $values)
+    {
+        $refClass = new ReflectionClass($this->className);
+        $constructor = $refClass->getConstructor();
+        $constructorParameters = $constructor ? $constructor->getParameters() : [];
+
+        $parameters = [];
+        foreach ($constructorParameters as $parameter) {
+            $name = $parameter->getName();
+            if (!$parameter->isDefaultValueAvailable() && empty($values[$name])) {
+                throw FailedResolvingInputType::createForMissingConstructorParameter($refClass->getName(), $name);
+            }
+
+            $parameters[] = $values[$name] ?? $parameter->getDefaultValue();
+        }
```

Comment:
I'm not sure if we should overload constructor with such logic. It's job is only to instantiate an object with a given parameters. Current implementation assumes that required constructor parameters have defined fields on properties with the same name in the class. When object is instantiated users payload is mapped to the constructor parameters.

For example:
```
/**
 * @GQL\Input()
 */
class User
{
    /**
     * @var string Username or login
     *
     * @GQL\Field()
     */
    private string $username;

    /**
     * @param string $username
     */
    public function __construct(string $username)
    {
        $this->username = $username;
    }
}
```

In this case we definitely know that input type has a field called `username` and it's required. There's a constructor parameter with the same name so when the object is instantiated value from the request payload will be passed to it. For this `createInstance()` function is responsible - this is kind of default factory, which knows how to map data to a constructor and instantiate an object.

I agree that it would be nice to provide a custom and flexible way of creating instances before they are populated with new data. At the beginning I thought of reusing `@Factory`. We could add a new attribute `@Input(init="FactoryName")`. But after that I realized that it's not very flexible. For example it would be nice to reuse the same input type in different mutations, but pre-load should be done with different identificators. For example: 
```
updateUser(id: ID!, user: UserInput!)
updateUserByEmail(email: String!, user: UserInput)
updateMe(user: UserInput)
```

Considering this we'd need to create 3 different input types with the same fields but different `init` attributes.

As an alternative this `init` attribute could be added to `@UseInputType` annotation. This way we choose what type should be used and how the object should be initialized for each particular mutation. Also extra parameters such as "id: ID!" or "email: String!" described in factories could be mapped not to input types, but to mutations directly. I hope it won't be confusing. Here's an example of what I mean:

```
/**
 * @GQL\Type()
 * @GQL\Input()
 */
class User
{

    /**
     * @var string
     *
     * @GQL\Field(for="User")
     */
    private string $id;
    
    /**
     * @var string
     *
     * @GQL\Field()
     */
    private string $username;

    ...
}

...

class UserFactory
{

    /**
     * @Factory(name="LoadUser")
     * 
     * @param string $id
     * 
     * @return User
     */
    public function loadUser(string $id): User {
       // ... Load user from DB
    }
}

...

class UserController
{

    /**
     * @Mutation(name="UpdateUser")
     * @UseInputType(for="$user", inputType="UserInput!", init="LoadUser")
     * 
     * @param User $user
     * 
     * @return User
     */
    public function updateUser(User $user): User {
       // ... $user is loaded from DB and populated with given data. Just validate and save it. 
    }
}
```

Because `@UseInputType` points to "LoadUser" factory with `init` attribute, another field `id` will be added to the mutation as `loadUser()` function requires it as a parameter. So the request mapping will be like this:
```
updateUserByEmail(id: String!, user: UserInput)
```

My only concern is that it might be not obvious that by adding `init` attribute your mutation signature is split across different methods - actual mutation method and supportive "init" factory. It might be a bit confusing or not obvious...
What do you think about this idea? Do you see other ways to achieve this?

Reply:
About the "init" idea: I would even go one step further. The concept of @factory is hard to explain and I'm getting a lot of people confused with it.

Since quite often, a factory is used by only one mutation, we could even imagine something like this:

```php
class UserController
{

    /**
     * @Mutation(name="UpdateUser")
     * @UseInputType(for="$user", factory="UserFactory::loadUser", inputName="UpdateUserInput")
     * 
     * @param User $user
     * 
     * @return User
     */
    public function updateUser(User $user): User {
       // ... $user is loaded from DB and populated with given data. Just validate and save it. 
    }
}
```

and this might actually be easier to understand for many people.

Reply:
Is `UserFactory` just some service with `loadUser` function and has nothing to do with `@Factory` annotation?

I considered similar approach, but in this case you'd need to specify an absolute class name. For example: `@UseInputType(for="$user", factory="App\Services\UserFactory::loadUser", inputName="UpdateUserInput")`. It will be easier to understand and use. Not sure only if it looks good in the code though :smile: